File: packages/deon-go/resources.go
package deon

import (
	"net/url"
	"os"
	"path"
	"strings"
	"unicode/utf8"
)

// Where a resource comes from, and whether it may be reached at all (specification 9). Nothing is
// granted by default: a target is refused *before* a request is made, not after one comes back, which
// is the whole difference between a decision and an accident. Two failures must never be confused —
// DEON_CAPABILITY_DENIED is a policy, DEON_RESOURCE_IO is the world.

type fetched struct {
	data       string
	filetype   string // the extension an import is read as; empty for an injection
	filebase   string // what a relative target inside the loaded document resolves against
	resourceID string // the canonical identity, which the cycle check compares
}

func isURL(target string) bool {
	parsed, err := url.Parse(target)
	return err == nil && (parsed.Scheme == "http" || parsed.Scheme == "https")
}

func extensionOf(target string) string {
	if isURL(target) {
		if parsed, err := url.Parse(target); err == nil {
			return path.Ext(parsed.Path)
		}
	}
	return path.Ext(target)
}

func directoryOf(target string) string {
	if isURL(target) {
		if parsed, err := url.Parse(target); err == nil {
			parsed.Path = path.Dir(parsed.Path) + "/"
			return parsed.String()
		}
	}
	return path.Dir(target)
}

// resolveTarget turns a target as written into the target it names (specification 9). A relative
// filesystem target resolves against the document's filebase; a relative URL against the document's
// URL. The absolutePaths mapping is applied last.
func (in *interpreter) resolveTarget(target string) string {
	resolved := target

	switch {
	case isURL(target):
		// already absolute
	case strings.HasPrefix(target, "/"):
		// a logical absolute path, left for the absolutePaths mapping
	case in.options.Filebase != "" && isURL(in.options.Filebase):
		base := in.options.Filebase
		if !strings.HasSuffix(base, "/") {
			base += "/"
		}
		if ref, err := url.Parse(target); err == nil {
			if root, err := url.Parse(base); err == nil {
				resolved = root.ResolveReference(ref).String()
			}
		}
	default:
		resolved = path.Join(in.options.Filebase, target)
	}

	return in.mapAbsolute(resolved)
}

// mapAbsolute applies the absolutePaths option: exact keys win before wildcards; among /prefix/*
// wildcards the longest prefix wins, and the unmatched suffix is appended to the mapped directory.
func (in *interpreter) mapAbsolute(target string) string {
	paths := in.options.AbsolutePaths
	if len(paths) == 0 {
		return target
	}
	if mapped, ok := paths[target]; ok {
		return mapped
	}

	best := ""
	for key := range paths {
		if !strings.HasSuffix(key, "/*") {
			continue
		}
		prefix := key[:len(key)-1] // keep the trailing slash
		if strings.HasPrefix(target, prefix) && len(prefix) > len(best) {
			best = prefix
		}
	}
	if best == "" {
		return target
	}
	mapped := paths[best+"*"]
	remainder := target[len(best):]
	return strings.TrimRight(mapped, "/") + "/" + remainder
}

// load fetches a resource, gating on the capability it needs. In-memory resources supplied to the
// evaluator are consulted first, so a document handed its own resources reaches neither a disk nor a
// network to find them.
func (in *interpreter) load(target, kind string, span Span) fetched {
	filetype := ""
	if kind == "import" {
		filetype = extensionOf(target)
	}

	if data, ok := in.options.Resources[target]; ok {
		return fetched{data: data, filetype: filetype, filebase: directoryOf(target), resourceID: target}
	}

	if isURL(target) {
		if !in.options.AllowNetwork {
			fail(CapabilityDenied,
				"The resource '"+target+"' was not permitted: network access is not allowed.", span)
		}
		data := in.fetchOverNetwork(target, kind, span)
		return fetched{data: data, filetype: filetype, filebase: directoryOf(target), resourceID: target}
	}

	if !in.options.AllowFilesystem {
		fail(CapabilityDenied,
			"The resource '"+target+"' was not permitted: filesystem access is not allowed.", span)
	}

	bytes, err := os.ReadFile(target)
	if err != nil {
		fail(ResourceIO, "Unable to read resource '"+target+"'.", span)
	}
	if !isValidUTF8(string(bytes)) {
		fail(ResourceFormat, "The resource '"+target+"' is not valid UTF-8.", span)
	}
	return fetched{data: string(bytes), filetype: filetype, filebase: directoryOf(target), resourceID: target}
}

// #region import and injection
func (in *interpreter) evalImport(decl *declaration) Value {
	token := in.resolveAuthenticator(decl)
	target := in.importTarget(in.resolveTarget(decl.target))

	f := in.load(target, "import", decl.span)
	_ = token

	if in.opened[f.resourceID] {
		// A resource that imports its way back to itself is a cycle, reported at the statement that
		// closed the loop (§11.2).
		fail(Cycle, "The resource '"+f.resourceID+"' imports itself.", decl.span)
	}
	in.opened[f.resourceID] = true
	defer delete(in.opened, f.resourceID)

	if f.filetype == ".json" {
		return in.jsonAtStatement(f.data, decl.span)
	}
	if f.filetype != "" && f.filetype != ".deon" {
		fail(ResourceFormat, "The import '"+decl.target+"' has an unsupported extension.", decl.span)
	}

	return in.importDeon(f, decl.span)
}

// importTarget appends the default extension: when an import target has no extension, `.deon` is
// appended (specification 9).
func (in *interpreter) importTarget(target string) string {
	if extensionOf(target) == "" {
		return target + ".deon"
	}
	return target
}

// importDeon parses and evaluates an imported Deon document, re-anchoring any diagnostic it raises to
// the importing statement (§11.2): the document a caller is holding is the importing one.
func (in *interpreter) importDeon(f fetched, at Span) Value {
	value, err := func() (v Value, err error) {
		defer recoverError(&err)
		doc := mustParse(f.data, f.resourceID)

		sub := newInterpreter(in.subOptions(f))
		sub.opened = in.opened
		return sub.run(doc), nil
	}()
	if err != nil {
		reanchor(err, at)
		panic(err)
	}
	return value
}

func (in *interpreter) evalInject(decl *declaration) Value {
	_ = in.resolveAuthenticator(decl)
	target := in.resolveTarget(decl.target)
	f := in.load(target, "inject", decl.span)
	return f.data
}

func (in *interpreter) subOptions(f fetched) *ParseOptions {
	sub := *in.options
	sub.SourceName = f.resourceID
	sub.Filebase = f.filebase
	return &sub
}

// resolveAuthenticator resolves an optional `with authenticator` to a string: a literal, an
// environment link, or a leaflink that resolves to a string.
func (in *interpreter) resolveAuthenticator(decl *declaration) string {
	if decl.authenticator == nil {
		return ""
	}
	value := in.eval(decl.authenticator)
	text, ok := value.(string)
	if !ok {
		// An authenticator that resolves to a map or list is the wrong shape, not a malformed resource:
		// nothing has been fetched yet. It is a type mismatch, reported at the importing statement.
		fail(TypeMismatch, "An authenticator must resolve to a string.", decl.span)
	}
	return text
}

// #endregion import and injection

// reanchor rewrites the primary span of a raised error to the importing statement, leaving the
// original location in the message for the import trace. A cycle keeps its own span.
func reanchor(err error, at Span) {
	deonErr, ok := err.(*Error)
	if !ok || len(deonErr.Diagnostics) == 0 {
		return
	}
	if deonErr.Code == Cycle {
		return
	}
	deonErr.Diagnostics[0].Span = at
}

func isValidUTF8(s string) bool { return utf8.ValidString(s) }

// mustParse parses a document and raises on failure, for use inside a boundary that will recover.
func mustParse(text, source string) *document {
	doc, err := parseSyntax(text, source)
	if err != nil {
		panic(err)
	}
	return doc
}

// jsonAtStatement converts imported JSON, re-anchoring a format fault to the importing statement.
func (in *interpreter) jsonAtStatement(data string, at Span) Value {
	value, err := func() (v Value, err error) {
		defer recoverError(&err)
		return jsonToValue(data, at), nil
	}()
	if err != nil {
		reanchor(err, at)
		panic(err)
	}
	return value
}

// fetchOverNetwork retrieves a resource over HTTP once the network has been granted. The conformance
// suite never reaches a public network (specification 15), so this exists for real use rather than for
// the fixtures, which supply their resources in memory.
func (in *interpreter) fetchOverNetwork(target, kind string, span Span) string {
	return httpGet(target, kind, in.tokenFor(target), span)
}

// tokenFor is the bearer credential for a host, from the authorization option keyed by exact
// lowercase hostname (specification 9).
func (in *interpreter) tokenFor(target string) string {
	parsed, err := url.Parse(target)
	if err != nil {
		return ""
	}
	if token, ok := in.options.Authorization[strings.ToLower(parsed.Hostname())]; ok {
		return token
	}
	return ""
}

File: packages/deon-rust/core/src/resources.rs
//! Where a resource lives, and whether it may be reached at all.

use std::collections::HashMap;

use crate::options::ParseOptions;
use crate::syntax::{Resource, ResourceKind};
use crate::text::{is_absolute_path, is_url, scheme_of};
use crate::url::Url;

/// What a loader hands back.
pub struct Fetched {
    /// The bytes as they were read. A loader answers *whether* a resource could be reached, not
    /// whether what it reached is text: bytes that are not valid UTF-8 were read successfully, so
    /// their encoding is a format fault the interpreter raises when it decodes them, not an I/O one
    /// the loader reports by handing back nothing (specification 1, 9).
    pub data: Vec<u8>,

    /// `.deon`, `.json`, or the empty string for an injection, which keeps its target exactly.
    pub filetype: String,

    /// What a relative target inside the loaded document resolves against.
    pub filebase: String,

    /// The canonical identity of what was actually read, which is what the cycle check compares.
    pub resource_id: String,
}

/// Reads a resource, or does not. A parser built without one cannot reach the filesystem or the
/// network by accident, because it has nothing to reach with (specification 9).
///
/// A host that wants a resource from somewhere this crate does not know about — a database, an
/// archive, an HTTP client of its own choosing — implements this, and needs no feature flag to do it.
///
/// `token` is the bearer credential already decided for this target: the `with` written on the
/// declaration if there is one, otherwise the `authorization` entry for the host. Deciding it is not
/// the loader's job; using it is.
pub trait ResourceLoader {
    fn load(
        &self,
        target: &str,
        kind: ResourceKind,
        options: &ParseOptions,
        token: Option<&str>,
    ) -> Option<Fetched>;
}

/// A loader that can reach nothing at all.
pub struct DenyAll;

impl ResourceLoader for DenyAll {
    fn load(
        &self,
        _target: &str,
        _kind: ResourceKind,
        _options: &ParseOptions,
        _token: Option<&str>,
    ) -> Option<Fetched> {
        None
    }
}

/// The filesystem, and only when it has been granted.
///
/// A remote target is not this loader's business, so it is refused by returning nothing — which the
/// interpreter turns into the denial or the failure that the capability model calls for. Reaching a
/// remote target is [`crate::network::Http`]'s job, and it exists only behind the `network` feature.
pub struct Filesystem;

impl ResourceLoader for Filesystem {
    fn load(
        &self,
        target: &str,
        kind: ResourceKind,
        options: &ParseOptions,
        _token: Option<&str>,
    ) -> Option<Fetched> {
        if is_url(target) || scheme_of(target).is_some() {
            return None;
        }

        if !options.allow_filesystem {
            return None;
        }

        let file = resolve_mapped_absolute_path(target, &options.absolute_paths);

        // The bytes, as they are. A file that is not there or may not be read is `None`, which the
        // interpreter turns into an I/O failure; a file that is there but is not UTF-8 is read
        // successfully here and judged a format fault when the interpreter decodes it (specification
        // 1, 9). Reading to a `String` would collapse the two into one indistinguishable `None`.
        let data = std::fs::read(&file).ok()?;

        Some(Fetched {
            data,
            filetype: extension(&file, kind),
            filebase: directory_of(&file).to_string(),
            resource_id: file,
        })
    }
}

/// Everything this crate itself knows how to reach: the filesystem, and — with the `network` feature
/// — a remote target.
///
/// Neither is granted here. Each sub-loader refuses on its own unless the matching capability was
/// asked for, so the capability model is enforced in one place per capability rather than at every
/// call site (specification 9).
pub struct Host;

impl ResourceLoader for Host {
    fn load(
        &self,
        target: &str,
        kind: ResourceKind,
        options: &ParseOptions,
        token: Option<&str>,
    ) -> Option<Fetched> {
        #[cfg(feature = "network")]
        if let Some(fetched) = crate::network::Http.load(target, kind, options, token) {
            return Some(fetched);
        }

        Filesystem.load(target, kind, options, token)
    }
}

/// Maps a logical absolute target onto the host path that actually holds it (specification 9).
///
/// An exact key wins before any wildcard. Among the wildcards, which end in `/*`, the longest prefix
/// wins, and whatever of the target the prefix did not match is appended to the mapped directory.
///
/// This is applied before the loader, rather than inside it, because the mapping is a property of
/// the target and not of whoever resolves it: a resource handed over through `resources` must map
/// exactly as one read from a disk, or the same document would mean two different things.
pub fn resolve_mapped_absolute_path(file: &str, mappings: &HashMap<String, String>) -> String {
    // Only an absolute target is logical. A relative one resolves against the document holding it.
    if !is_absolute_path(file) {
        return file.to_string();
    }

    if let Some(mapped) = mappings.get(file) {
        return mapped.clone();
    }

    let wildcard = mappings
        .keys()
        .filter(|key| key.ends_with("/*") && file.starts_with(&key[..key.len() - 1]))
        .max_by_key(|key| key.len());

    let Some(wildcard) = wildcard else {
        return file.to_string();
    };

    let prefix = &wildcard[..wildcard.len() - 1];
    let suffix = &file[prefix.len()..];
    let directory = mappings[wildcard].trim_end_matches('/');

    format!("{directory}/{suffix}")
}

/// The canonical identity of a resource. A relative filesystem target resolves against the file that
/// holds it, a relative URL against the URL that holds it, and `..` is folded away so that two
/// spellings of one resource cannot escape the cycle check.
pub fn resource_target(resource: &Resource, options: &ParseOptions) -> String {
    let target = with_import_extension(&resource.target, resource.kind);

    // A target with a scheme of its own is not relative to anything.
    if scheme_of(&target).is_some() {
        return match Url::parse(&target) {
            Some(url) => url.href(),
            None => target,
        };
    }

    let source = &options.source_name;
    let filebase = &options.filebase;

    // A document read over HTTP resolves its own relative imports over HTTP.
    //
    // The asymmetry is deliberate, and is the reference's: a `filebase` is a *directory*, so it is
    // given a trailing slash and the reference is appended to it, whereas a `source_name` is the
    // *document* itself, so the reference replaces its last segment.
    let url_base = if is_url(filebase) {
        Some(format!("{}/", filebase.trim_end_matches('/')))
    } else if is_url(source) {
        Some(source.clone())
    } else {
        None
    };

    if let Some(url_base) = url_base {
        if let Some(base) = Url::parse(&url_base) {
            return base.join(&target).href();
        }
    }

    let normalized = target.replace('\\', "/");
    let absolute = is_absolute_path(&normalized);

    let base = if !filebase.is_empty() {
        filebase.as_str()
    } else {
        directory_of(source)
    };

    // An absolute target stands on its own, and a relative one with nothing to stand against is
    // taken as written.
    let joined = if absolute || base.is_empty() {
        normalized
    } else {
        format!("{base}/{normalized}")
    };

    normalize_path(&joined)
}

/// The `authorization` option is keyed by an exact lowercase hostname — no port, no path, no
/// wildcard. A target that is not remote gets no token at all.
pub fn authorization(target: &str, options: &ParseOptions) -> Option<String> {
    if !is_url(target) {
        return None;
    }

    let url = Url::parse(target)?;

    options.authorization.get(url.hostname()).cloned()
}

/// Folds `.` and `..` away. Above the root there is nothing, so a rooted path drops what it cannot
/// climb.
fn normalize_path(joined: &str) -> String {
    let drive = match joined.as_bytes() {
        [letter, b':', ..] if letter.is_ascii_alphabetic() => &joined[..2],
        _ => "",
    };

    let rest = &joined[drive.len()..];
    let rooted = rest.starts_with('/') || !drive.is_empty();

    let mut normalized: Vec<&str> = Vec::new();

    for part in rest.split('/') {
        match part {
            "" | "." => continue,
            ".." => {
                if normalized.last().is_some_and(|last| *last != "..") {
                    normalized.pop();
                } else if !rooted {
                    normalized.push(part);
                }
            }
            _ => normalized.push(part),
        }
    }

    let root = if rest.starts_with('/') { "/" } else { "" };

    format!("{drive}{root}{}", normalized.join("/"))
}

/// An import target with no extension is a Deon document. The query and the fragment are not part of
/// the path, so the extension goes before them.
pub fn with_import_extension(target: &str, kind: ResourceKind) -> String {
    if kind != ResourceKind::Import {
        return target.to_string();
    }

    let suffix_at = target.find(['?', '#']);

    let (pathname, suffix) = match suffix_at {
        Some(at) => (&target[..at], &target[at..]),
        None => (target, ""),
    };

    let slash = pathname
        .rfind(['/', '\\'])
        .map(|at| at as isize)
        .unwrap_or(-1);

    let dot = pathname.rfind('.').map(|at| at as isize).unwrap_or(-1);

    if dot > slash {
        return target.to_string();
    }

    format!("{pathname}.deon{suffix}")
}

/// An injection keeps its target exactly, without parsing it, so it has no format at all.
pub fn extension(target: &str, kind: ResourceKind) -> String {
    if kind == ResourceKind::Inject {
        return String::new();
    }

    let clean = match target.find(['?', '#']) {
        Some(at) => &target[..at],
        None => target,
    };

    let slash = clean.rfind('/').map(|at| at as isize).unwrap_or(-1);
    let dot = clean.rfind('.').map(|at| at as isize).unwrap_or(-1);

    if dot > slash {
        clean[dot as usize..].to_lowercase()
    } else {
        ".deon".to_string()
    }
}

pub fn directory_of(path: &str) -> &str {
    match path.rfind('/') {
        Some(at) => &path[..at],
        None => "",
    }
}
